git-hooks/enforce-workflow/enforce-workflow.rb

```diff
@@ -0,0 +1,316 @@
+        puts '*** Creating a branch is not allowed in this repository.'
+        update_allowed = false
+        break
+      end
+
+      if REPO.config['hooks.allowunsignedcommits'] != 'true'
+        if !signed
+          puts "*** Bad signature on commit #{commit.oid}."
+          update_allowed = false
+          break
+        elsif signer
+          puts "*** Good signature on commit #{commit.oid} by #{signer} (#{fingerprint})."
+        else
+          # Signed, but not allowed
+          puts "*** Commit #{commit.oid} signed by unauthorised key #{fingerprint}."
+          update_allowed = false
+          break
+        end
+      end
+
+    when :merge
+      if REPO.config['hooks.allowunsignedcommits'] != 'true'
+        if !signed
+          puts "*** Bad signature on merge #{commit.oid}."
+          update_allowed = false
+          break
+        elsif signer
+          puts "*** Good signature on merge #{commit.oid} by #{signer} (#{fingerprint})."
+        else
+          # Signed, but not allowed
+          puts "*** Merge #{commit.oid} signed by unauthorised key #{fingerprint}."
+          update_allowed = false
+          break
+        end
+      end
+
+    else
+      puts "*** Unknown type of update to ref #{ref} of type #{commit_type}."
+      update_allowed = false
+      break
+    end
+  end
+
+  return [update_allowed, commit_count]
+end
+
+# Check if the lightweight tag +ref+ should be allowed to be created.
+def allow_lightweight_tag?(ref)
+  if (REPO.config['hooks.allowunsignedtags'] != 'true') ||
+     (REPO.config['hooks.allowunannotated'] != 'true')
+    puts "*** The un-annotated tag #{ref} is not allowed in this repository."
+    puts "*** Use 'git tag [ -a | -s ]' for tags you want to propagate."
+    return false
+  end
+  return true
+end
+
+# Check if the annotated tag +ref+, currently pointing to commit ID
+# +old+, should be allowed to be created or updated to point to commit
+# ID +new+.
+def allow_annotated_tag?(old, new, ref)
+  if old.to_i(16).nonzero? && (REPO.config['hooks.allowmodifytag'] != 'true')
+    puts "*** Tag #{ref} already exists."
+    puts '*** Modifying a tag is not allowed in this repository.'
+    return false
+  elsif REPO.config['hooks.allowunsignedtags'] != 'true'
+    signed = false
+    fingerprint = ''
+    signer = ''
+    signature, plaintext = Rugged::Tag.extract_signature(REPO, new)
+    if signature && plaintext
+      CRYPTO.verify(signature, signed_text: plaintext) do |signature|
+        signed = signature.valid?
+        if signed
+          fingerprint = signature.fingerprint
+          signer = find_signer(fingerprint)
+        end
+      end
+    end
+    if !signed
+      puts "*** Bad signature on tag #{ref}."
+      return false
+    elsif signer
+      puts "*** Good signature on tag #{ref} by #{signer} (#{fingerprint})."
+      return true
+    else
+      puts "*** Rejecting tag #{ref} due to lack of a valid GPG signature."
```

Comment:
Lies! It has a valid signature, but the signer is not authorized (see `walk_graph`). 

Also, maybe consider a DRYer approach to signature validation and reporting?